# 1. EDA 과정

### 1-1 목표
부분유료화 게임, 유료게임의 차이점과  
유료 -> 부분유료화의 시점을확인.  
이때 인터넷방송 시청자와 관계가 있는지 확인.  
부분유료화의 시점은 언제가 적당한지 확인.  
대형패치의 적절한 시점 확인.  
5개의 게임 확인후 적절한 데이터셋 지정하기. 

In [2]:
import pandas as pd
import numpy as np

In [1]:
CounterStrike_Global_Offensive = pd.read_csv('/content/Lifetime player count history CS GO.csv')
Apex_Legends = pd.read_csv('/content/Lifetime player count history APEX.csv')
PUBG_BATTLEGROUNDS = pd.read_csv('/content/Lifetime player count history battleground.csv')
Tom_Clancys_Rainbow_Six_Siege = pd.read_csv('/content/Lifetime player count history rainbow six siege.csv')
Team_Fortress2 = pd.read_csv('/content/Lifetime player count history team fortress2.csv')

In [2]:
# 특성공학.
def engineering(df):
  index = df.DateTime.str.contains('-').index
  # 결측치는 0으로 삽입
  df.fillna(0, inplace=True)
  # DateTime 뒤 시간은 없애고 날짜만 추출하기위한 for문
  for i in index:
    df['DateTime'].iloc[i] = df['DateTime'].iloc[i][:10]
  # float type int 타입으로 변환
  for feature in ['Players', 'Average Players', 'Twitch Viewers', 'Flags']:
    df[feature] = df[feature].astype(int)

  return df

# 함수 적용
engineering(CounterStrike_Global_Offensive)
engineering(Apex_Legends)
engineering(PUBG_BATTLEGROUNDS)
engineering(Tom_Clancys_Rainbow_Six_Siege)
engineering(Team_Fortress2)

<ipython-input-2-dd0fd3352062>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['DateTime'].iloc[i] = df['DateTime'].iloc[i][:10]


DateTime  Players  Average Players  Twitch Viewers  Flags
0     2007-09-19     9553                0               0      0
1     2007-09-20        0                0               0      0
2     2007-09-21     5118                0               0      0
3     2007-09-22     4616                0               0      0
4     2007-09-23        0                0               0      0
...          ...      ...              ...             ...    ...
5677  2023-04-05    90031            83688             506      0
5678  2023-04-06    99494            84556             461      0
5679  2023-04-07   100103            89729             541      0
5680  2023-04-08    97511            91035            1630      0
5681  2023-04-09    98221            91035             884      0

[5682 rows x 5 columns]

In [3]:
# 게임명 컬럼 생성
CounterStrike_Global_Offensive['Title'] = np.repeat('CounterStrike_Global_Offensive', len(CounterStrike_Global_Offensive))
Apex_Legends['Title'] = np.repeat('Apex_Legends', len(Apex_Legends))
PUBG_BATTLEGROUNDS['Title'] = np.repeat('PUBG_BATTLEGROUNDS', len(PUBG_BATTLEGROUNDS))
Tom_Clancys_Rainbow_Six_Siege['Title'] = np.repeat('Tom_Clancys_Rainbow_Six_Siege', len(Tom_Clancys_Rainbow_Six_Siege))
Team_Fortress2['Title'] = np.repeat('Team_Fortress2', len(Team_Fortress2))

In [26]:
# dlc발매 or 큰 이벤트 컬럼 수정
dfindex = CounterStrike_Global_Offensive[CounterStrike_Global_Offensive['DateTime'] == '2021-09-22'].index
CounterStrike_Global_Offensive['Flags'][dfindex] = 'Operation Riptide'
CounterStrike_Global_Offensive.iloc[dfindex]

<ipython-input-26-c4c3eecfe5d5>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  CounterStrike_Global_Offensive['Flags'][dfindex] = 'Operation Riptide'


DateTime  Players  Average Players  Twitch Viewers              Flags  \
3584  2021-09-22   949817                0          288058  Operation Riptide   

                               Title  
3584  CounterStrike_Global_Offensive

In [36]:
# dlc발매 or 큰 이벤트 컬럼 수정
dfindex = Apex_Legends[Apex_Legends['DateTime'] == '2023-02-14'].index
Apex_Legends['Flags'][dfindex] = 'Revelry'
Apex_Legends.iloc[dfindex]

<ipython-input-36-0451624e16ae>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Apex_Legends['Flags'][dfindex] = 'Revelry'


DateTime  Players  Average Players  Twitch Viewers    Flags  \
1471  2023-02-14   390627           208396          205164  Revelry   

             Title  
1471  Apex_Legends

In [43]:
# dlc발매 or 큰 이벤트 컬럼 수정
dfindex = PUBG_BATTLEGROUNDS[PUBG_BATTLEGROUNDS['DateTime'] == '2022-01-11'].index
PUBG_BATTLEGROUNDS['Flags'][dfindex] = 'Free Edition released'
PUBG_BATTLEGROUNDS.iloc[dfindex]

<ipython-input-43-604e7a0672c1>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  PUBG_BATTLEGROUNDS['Flags'][dfindex] = 'Free Edition released'


DateTime  Players  Average Players  Twitch Viewers  \
1950  2022-01-11    22055                0            3067   

                      Flags               Title  
1950  Free Edition released  PUBG_BATTLEGROUNDS

In [62]:
# dlc발매 or 큰 이벤트 컬럼 수정
dfindex = Tom_Clancys_Rainbow_Six_Siege[Tom_Clancys_Rainbow_Six_Siege['DateTime'] == '2022-12-01'].index
Tom_Clancys_Rainbow_Six_Siege['Flags'][dfindex] = 'Free Weekend'
Tom_Clancys_Rainbow_Six_Siege.iloc[dfindex]

<ipython-input-62-fc7f9a92cf42>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Tom_Clancys_Rainbow_Six_Siege['Flags'][dfindex] = 'Free Weekend'


DateTime  Players  Average Players  Twitch Viewers         Flags  \
2748  2022-12-01    36436            23736            6450  Free Weekend   

                              Title  
2748  Tom_Clancys_Rainbow_Six_Siege

In [69]:
# 중복행 삭제 및 인덱스 정렬
Tom_Clancys_Rainbow_Six_Siege.drop(index=1747, inplace=True)
Tom_Clancys_Rainbow_Six_Siege.reset_index(drop=True, inplace=True)

In [73]:
# dlc발매 or 큰 이벤트 컬럼 수정
dfindex = Team_Fortress2[Team_Fortress2['DateTime'] == '2019-07-26'].index
Team_Fortress2['Flags'][dfindex] = 'The Crate Depression of 2019'
Team_Fortress2.iloc[dfindex]

<ipython-input-73-d052d90e7d09>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Team_Fortress2['Flags'][dfindex] = 'The Crate Depression of 2019'


DateTime  Players  Average Players  Twitch Viewers  \
4328  2019-07-26    93245                0            3169   

                             Flags           Title  
4328  The Crate Depression of 2019  Team_Fortress2

In [76]:
# EDA 끝난 파일 내보내기
CounterStrike_Global_Offensive.to_csv('/content/eda/CounterStrike_Global_Offensive.csv', index=False)
Apex_Legends.to_csv('/content/eda/Apex_Legends.csv', index=False)
PUBG_BATTLEGROUNDS.to_csv('/content/eda/PUBG_BATTLEGROUNDS.csv', index=False)
Tom_Clancys_Rainbow_Six_Siege.to_csv('/content/eda/Tom_Clancys_Rainbow_Six_Siege.csv', index=False)
Team_Fortress2.to_csv('/content/eda/Team_Fortress2.csv', index=False)

In [106]:
data = pd.merge(CounterStrike_Global_Offensive, Apex_Legends, how='outer')
data = pd.merge(data, PUBG_BATTLEGROUNDS, how='outer')
data = pd.merge(data, Tom_Clancys_Rainbow_Six_Siege, how='outer')
data = pd.merge(data, Team_Fortress2, how='outer')
data

DateTime  Players  Average Players  Twitch Viewers Flags  \
0      2011-11-30      680                0               0     0   
1      2011-12-01        0                0               0     0   
2      2011-12-02        0                0               0     0   
3      2011-12-03        0                0               0     0   
4      2011-12-04        0                0               0     0   
...           ...      ...              ...             ...   ...   
16633  2023-04-05    90031            83688             506     0   
16634  2023-04-06    99494            84556             461     0   
16635  2023-04-07   100103            89729             541     0   
16636  2023-04-08    97511            91035            1630     0   
16637  2023-04-09    98221            91035             884     0   

                                Title  
0      CounterStrike_Global_Offensive  
1      CounterStrike_Global_Offensive  
2      CounterStrike_Global_Offensive  
3      CounterStrike_Global_Offensive  
4      CounterStrike_Global_Offensive  
...                               ...  
16633                  Team_Fortress2  
16634                  Team_Fortress2  
16635                  Team_Fortress2  
16636                  Team_Fortress2  
16637                  Team_Fortress2  

[16638 rows x 6 columns]

In [113]:
data['Title'].value_counts()

Team_Fortress2                    5682
CounterStrike_Global_Offensive    4149
Tom_Clancys_Rainbow_Six_Siege     2877
PUBG_BATTLEGROUNDS                2404
Apex_Legends                      1526
Name: Title, dtype: int64

In [112]:
data.isna().sum()

DateTime           0
Players            0
Average Players    0
Twitch Viewers     0
Flags              0
Title              0
dtype: int64

In [107]:
data.to_csv('/content/eda/data.csv', index=False)

In [116]:
data['DateTime'].duplicated().sum()

10956

In [114]:
data['DateTime'].sort_values()

10956    2007-09-19
10957    2007-09-20
10958    2007-09-21
10959    2007-09-22
10960    2007-09-23
            ...    
8078     2023-04-09
4148     2023-04-09
10955    2023-04-09
5674     2023-04-09
16637    2023-04-09
Name: DateTime, Length: 16638, dtype: object

In [ ]:
# 중복치는 게임 타이틀마다 해당 일 동시접속자수, 시청자수 때문에 따로 처리는 하지않음
# 가장긴 게임 날짜는 2007-09-19 ~ 2023-04-09, 5681일
# 중간에 flag의 전처리의 경우 직접 넣었기 때문에 추가셀은 하지않고 한셀에서 반복작업함.

# 2. 시각화

In [1]:
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [5]:
data = pd.read_csv('/content/data.csv')
apex = pd.read_csv('/content/Apex_Legends.csv')
csgo = pd.read_csv('/content/CounterStrike_Global_Offensive.csv')
pubg = pd.read_csv('/content/PUBG_BATTLEGROUNDS.csv')
team = pd.read_csv('/content/Team_Fortress2.csv')
tom = pd.read_csv('/content/Tom_Clancys_Rainbow_Six_Siege.csv')

In [6]:
# 평균 접속자수는 필요없는 컬럼이므로 제거.
for df in [apex, csgo, pubg, team, tom]:
  df.drop(columns='Average Players', inplace=True)
  df['Flags'].replace('0', np.nan, inplace=True)

In [7]:
# DLC or 무료화 지점 확인하기위한 셀
team[team['Flags'].notnull()]

DateTime  Players  Twitch Viewers                         Flags  \
21    2007-10-10     6020               0                 Steam Release   
1373  2011-06-23    71677               0         Free Edition released   
4110  2018-12-20   122938           10381                  Bot activity   
4328  2019-07-26    93245            3169  The Crate Depression of 2019   

               Title  
21    Team_Fortress2  
1373  Team_Fortress2  
4110  Team_Fortress2  
4328  Team_Fortress2

In [8]:
import plotly.graph_objects as go

# 시각화 함수 정의
def plot(df):
  dfflag = df[df['Flags'].notnull()]
  fig = go.Figure()
  fig.add_trace(go.Scatter(x=df['DateTime'],
                           y=df['Players'],
                           mode='lines',
                           name='Players',
                           opacity=0.5
                           ))
  fig.add_trace(go.Scatter(x=df['DateTime'],
                           y=df['Twitch Viewers'],
                           mode='lines',
                           name='Viewers',
                           opacity=0.2
                           ))
  fig.update_layout(
      title=df['Title'][0],
      template='plotly_white',)
  
  for date, pl, flag in zip(dfflag['DateTime'], dfflag['Players'], dfflag['Flags']):
    fig.add_annotation(x=date,
                       y=pl,
                       text=flag,)
  fig.show()

In [9]:
plot(apex)
plot(csgo)
plot(pubg)
plot(team)
plot(tom)

### 2-1 가설
데이터 EDA 결과 
대형패치의 가장 적절한 시점은 3개월~12개월 사이  
유료 -> 부분유료화의 경우 접속자수가 현저히 떨어져갈때 1년안에 결정을 지어야한다.  
인터넷방송 시청자와 접속자의 관계는 크게 있지않다.  

연속적인 데이터이므로 회귀문제이며.  
baseline은 평균값으로 할것.  
모델은 XGBRegressor 지정.  
데이터셋은 CounterStrike_Global_Offensive로 지정.  

# 모델링

In [44]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from xgboost import XGBRegressor
from sklearn.impute import SimpleImputer
from sklearn.pipeline import make_pipeline
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.metrics import mean_absolute_error, mean_squared_log_error
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from category_encoders import OneHotEncoder
from sklearn.model_selection import cross_val_score

In [ ]:
# 오류나면실행
# import warnings
# warnings.filterwarnings(action='ignore')

In [10]:
# 프로 경기 데이터셋 추가 매치 날짜만 추출해서 기존 csgo 데이터셋에 병합
csgomatch = pd.read_csv('/content/csgo_games.csv')
csgomatch

match_date       team_1       team_2  t1_points  t2_points  \
0     2016-12-18       CLOUD9  HELLRAISERS         13         16   
1     2016-12-18    IMMORTALS           G2         17         19   
2     2016-12-18  MOUSESPORTS    IMMORTALS         16          3   
3     2016-12-18     DIGNITAS           G2         16          9   
4     2016-12-18        OPTIC  HELLRAISERS         16         10   
...          ...          ...          ...        ...        ...   
3782  2020-01-25   VIRTUS.PRO       CLOUD9          0          2   
3783  2020-01-24       HEROIC    MAD LIONS         10         16   
3784  2020-01-19       HEROIC        FORZE          0          2   
3785  2020-01-19        CR4ZY        FORZE          0          2   
3786  2020-01-10       CLOUD9         MIBR          3          1   

      t1_world_rank  t2_world_rank  t1_h2h_win_perc  t2_h2h_win_perc winner  \
0                 9             20         0.500000         0.500000     t2   
1                13             10         0.500000         0.500000     t2   
2                12             13         0.500000         0.500000     t1   
3                 6             10         0.416667         0.583333     t1   
4                 4             20         0.500000         0.500000     t1   
...             ...            ...              ...              ...    ...   
3782             19             17         0.500000         0.500000     t2   
3783             16             18         0.333333         0.666667     t2   
3784             16             13         0.857143         0.142857     t2   
3785             20             13         0.538462         0.461538     t2   
3786             18             14         0.400000         0.600000     t1   

      ...  t2_player5_dpr  t2_player5_spr  t2_player5_opk_ratio  \
0     ...            0.63            0.10                  1.05   
1     ...            0.69            0.09                  0.85   
2     ...            0.67            0.07                  0.79   
3     ...            0.69            0.09                  0.85   
4     ...            0.63            0.10                  1.05   
...   ...             ...             ...                   ...   
3782  ...            0.67            0.11                  0.80   
3783  ...            0.65            0.08                  0.64   
3784  ...            0.63            0.09                  0.89   
3785  ...            0.63            0.09                  0.89   
3786  ...            0.65            0.06                  1.06   

      t2_player5_opk_rating  t2_player5_wins_perc_after_fk  \
0                      0.92                          0.733   
1                      0.96                          0.739   
2                      0.94                          0.723   
3                      0.96                          0.739   
4                      0.92                          0.733   
...                     ...                            ...   
3782                   0.95                          0.681   
3783                   0.80                          0.811   
3784                   0.92                          0.747   
3785                   0.92                          0.747   
3786                   0.99                          0.702   

      t2_player5_fk_perc_in_wins  t2_player5_multikill_perc  \
0                          0.104                   0.147018   
1                          0.147                   0.135810   
2                          0.129                   0.148257   
3                          0.147                   0.135810   
4                          0.104                   0.147018   
...                          ...                        ...   
3782                       0.121                   0.130481   
3783                       0.082                   0.095975   
3784                       0.113                   0.154618   
3785                       0.113                   0.154618 

In [11]:
match_da = csgomatch['match_date'].to_frame()
match_da['DateTime'] = match_da['match_date']
match_da.columns = ['DateTime', 'match_date']
match_da.duplicated().sum()

2662

In [12]:
csgo_steam = csgo.merge(match_da, how='outer', on='DateTime')
csgo_steam.drop_duplicates(inplace=True, ignore_index=True)
csgo_steam

DateTime  Players  Twitch Viewers Flags  \
0     2011-11-30      680               0   NaN   
1     2011-12-01        0               0   NaN   
2     2011-12-02        0               0   NaN   
3     2011-12-03        0               0   NaN   
4     2011-12-04        0               0   NaN   
...          ...      ...             ...   ...   
4144  2023-04-05  1381271          188879   NaN   
4145  2023-04-06  1419084          241831   NaN   
4146  2023-04-07  1459519          346816   NaN   
4147  2023-04-08  1448795          328263   NaN   
4148  2023-04-09  1393830          280853   NaN   

                               Title match_date  
0     CounterStrike_Global_Offensive        NaN  
1     CounterStrike_Global_Offensive        NaN  
2     CounterStrike_Global_Offensive        NaN  
3     CounterStrike_Global_Offensive        NaN  
4     CounterStrike_Global_Offensive        NaN  
...                              ...        ...  
4144  CounterStrike_Global_Offensive        NaN  
4145  CounterStrike_Global_Offensive        NaN  
4146  CounterStrike_Global_Offensive        NaN  
4147  CounterStrike_Global_Offensive        NaN  
4148  CounterStrike_Global_Offensive        NaN  

[4149 rows x 6 columns]

In [13]:
csgo_steam['DateTime'] = csgo_steam.DateTime.apply(pd.to_datetime)
csgo_steam['match_date'] = csgo_steam.match_date.apply(pd.to_datetime)

In [14]:
# 혹시모르니 csv파일 저장
csgo_steam.to_csv('/content/csgo_steam.csv', index=False)

In [15]:
def ploter(df):
  dfflag = df[df['Flags'].notnull()]
  dfmatch = df[df['match_date'].notnull()]
  fig = go.Figure()
  fig.add_trace(go.Scatter(x=df['DateTime'],
                           y=df['Players'],
                           mode='lines',
                           name='Players',
                           opacity=0.5
                           ))
  fig.add_trace(go.Scatter(x=df['DateTime'],
                           y=df['Twitch Viewers'],
                           mode='lines',
                           name='Viewers',
                           opacity=0.2
                           ))
  fig.add_trace(go.Scatter(x=dfmatch['match_date'],
                           mode='markers'
                           ))
  fig.update_layout(
      title=df['Title'][0],
      template='plotly_white',)
  
  for date, pl, flag in zip(dfflag['DateTime'], dfflag['Players'], dfflag['Flags']):
    fig.add_annotation(x=date,
                       y=pl,
                       text=flag,)
  fig.show()
# 프로경기를 추가한 그래프 시각화
ploter(csgo_steam)

In [16]:
# 머신러닝에 맞게 특성공학 실행.
csgo_steam['match_date'] = csgo_steam['match_date'].notnull()
csgo_steam['Flags'] = csgo_steam['Flags'].notnull()
csgo_steam.drop(columns=['DateTime', 'Title'], inplace=True)
csgo_steam

Players  Twitch Viewers  Flags  match_date
0         680               0  False       False
1           0               0  False       False
2           0               0  False       False
3           0               0  False       False
4           0               0  False       False
...       ...             ...    ...         ...
4144  1381271          188879  False       False
4145  1419084          241831  False       False
4146  1459519          346816  False       False
4147  1448795          328263  False       False
4148  1393830          280853  False       False

[4149 rows x 4 columns]

In [80]:
# 대형패치(이벤트), 방송시청자수, 프로경기에 관해 게임 동시접속자수를 예측하는 모델을 목표.
# 타겟, 검증셋, 테스트셋 구분
target = 'Players'
X = csgo_steam.drop(columns=target)
y = csgo_steam[target]

X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_val, y_val, random_state=42)

X_train.shape, X_val.shape, X_test.shape, y_train.shape, y_val.shape, y_test.shape

((3111, 3), (778, 3), (260, 3), (3111,), (778,), (260,))

In [81]:
# 기준모델 지정(평균 접속자수)
baseline = [y_train.mean()] * len(y_train)

# 모델 생성
pipe = make_pipeline(
    OneHotEncoder(),
    SimpleImputer(),
    StandardScaler(),
    XGBRegressor(random_state=42)
)

pipe.fit(X_train, y_train)

Pipeline(steps=[('onehotencoder', OneHotEncoder(cols=[])),
                ('simpleimputer', SimpleImputer()),
                ('standardscaler', StandardScaler()),
                ('xgbregressor',
                 XGBRegressor(base_score=0.5, booster='gbtree', callbacks=None,
                              colsample_bylevel=1, colsample_bynode=1,
                              colsample_bytree=1, early_stopping_rounds=None,
                              enable_categorical=False, eval_metric=None,
                              feature_types=None, gamma=0, gpu_id=-1,
                              grow_policy='depthwise', importance_type=None,
                              interaction_constraints='',
                              learning_rate=0.300000012, max_bin=256,
                              max_cat_threshold=64, max_cat_to_onehot=4,
                              max_delta_step=0, max_depth=6, max_leaves=0,
                              min_child_weight=1, missing=nan,
                              monotone_constraints='()', n_estimators=100,
                              n_jobs=0, num_parallel_tree=1, predictor='auto',
                              random_state=42, ...))])

In [82]:
# R2_score , 기준모델 성능 확인
print('r2_score_train :', pipe.score(X_train, y_train))
print('r2_score_val :', pipe.score(X_val, y_val))
print('r2_score_baseline :', r2_score(y_train, baseline))
print('MAE_baseline :', mean_absolute_error(y_train, baseline))

r2_score_train : 0.8742165504037471
r2_score_val : 0.7569895661586306
r2_score_baseline : 0.0
MAE_baseline : 276328.14792441786


In [83]:
# 학습, 검증 ,테스트셋 함수 지정
def print_score(model, X_train, y_train, X_test, y_test) :

    train_score = np.round(model.score(X_train, y_train) , 3)
    val_score = np.round(np.mean(cross_val_score(model, X_train, y_train, scoring='r2', cv=3).round(3)),3)
    test_score = np.round(model.score(X_test, y_test),3)
    print(f'학습 세트 r2_score : {train_score}')
    print(f'검증 세트 r2_score : {val_score}')
    print(f'테스트 세트 r2_score : {test_score}')

    return train_score, val_score, test_score

In [79]:
# 선형회귀 모델 생성
poly_model = make_pipeline(
    SimpleImputer(),
    StandardScaler(),
    PolynomialRegression(2)
)
# 모델 학습
poly_model.fit(X_train, y_train)
# 성능 확인
ols_train, ols_val, ols_test = print_score(poly_model,X_train, y_train, X_test, y_test)

학습 세트 r2_score : 0.587
검증 세트 r2_score : 0.529
테스트 세트 r2_score : 0.56


XGBRegressor 성능이 더 높음으로 파라미터 조정

In [128]:
# 기준모델 지정(평균 접속자수)
baseline = [y_train.mean()] * len(y_train)

# 모델 생성
model = make_pipeline(
    SimpleImputer(),
    StandardScaler(),
    XGBRegressor(random_state=42,
                 max_depth=3)
)

model.fit(X_train, y_train)

Pipeline(steps=[('simpleimputer', SimpleImputer()),
                ('standardscaler', StandardScaler()),
                ('xgbregressor',
                 XGBRegressor(base_score=0.5, booster='gbtree', callbacks=None,
                              colsample_bylevel=1, colsample_bynode=1,
                              colsample_bytree=1, early_stopping_rounds=None,
                              enable_categorical=False, eval_metric=None,
                              feature_types=None, gamma=0, gpu_id=-1,
                              grow_policy='depthwise', importance_type=None,
                              interaction_constraints='',
                              learning_rate=0.300000012, max_bin=256,
                              max_cat_threshold=64, max_cat_to_onehot=4,
                              max_delta_step=0, max_depth=3, max_leaves=0,
                              min_child_weight=1, missing=nan,
                              monotone_constraints='()', n_estimators=100,
                              n_jobs=0, num_parallel_tree=1, predictor='auto',
                              random_state=42, ...))])

In [129]:
# 성능 확인
ols_train, ols_val, ols_test = print_score(model,X_train, y_train, X_test, y_test)

학습 세트 r2_score : 0.821
검증 세트 r2_score : 0.762
테스트 세트 r2_score : 0.772


2  
학습 세트 r2_score : 0.801  
검증 세트 r2_score : 0.77  
테스트 세트 r2_score : 0.78  
3  
학습 세트 r2_score : 0.821  
검증 세트 r2_score : 0.762   
테스트 세트 r2_score : 0.772  
4  
학습 세트 r2_score : 0.84  
검증 세트 r2_score : 0.754  
테스트 세트 r2_score : 0.761  
5  
학습 세트 r2_score : 0.856  
검증 세트 r2_score : 0.74  
테스트 세트 r2_score : 0.751  
6  
학습 세트 r2_score : 0.874  
검증 세트 r2_score : 0.732  
테스트 세트 r2_score : 0.743  

# 결론
1. 베이스라인보다 높게나왔다.  
베이스라인검증셋 : 0.756  
모델 검증셋 : 0.762  
2. 파라미터를 조정하였습니다. 데이터전처리의 경우 더 이상 건드릴게 없어보였기 떄문입니다.  
3. 최종모델인 XGBRegressor는 과적합을 예방함, 일반화시키며
이전 트리들에서의 에러를 수정해가며 회귀문제를 해결한 모델입니다.  
4. 	Players, Twitch Viewers, Flags, match_date



---


따라서 초기에 설정했던 가설중 한가지만 해결..   
~~데이터 EDA 결과 대형패치의 가장 적절한 시점은 3개월,12개월 사이~~  
~~유료, 부분유료화의 경우 접속자수가 현저히 떨어져갈때 1년안에 결정을 지어야한다.~~  
인터넷방송 시청자와 접속자의 관계는 상당히 높다고 볼수있다.